quickwit/quickwit-cli/src/main.rs

```diff
@@ -113,50 +113,106 @@ fn setup_logging_and_tracing(
+            };
+            quickwit_telemetry::send_telemetry_event(TelemetryEvent::EndCommand { return_code })
+                .await;
+            telemetry_handle.terminate_telemetry().await;
+            global::shutdown_tracer_provider();
+            std::process::exit(return_code)
+        })
+}
 
-    let app = build_cli()
-        .about(about_text.as_str())
-        .version(version_text.as_str());
-    let matches = app.get_matches();
+mod busy_detector {
+    use std::sync::atomic::{AtomicU64, Ordering};
+    use std::time::Instant;
 
-    let command = match CliCommand::parse_cli_args(&matches) {
-        Ok(command) => command,
-        Err(err) => {
-            eprintln!("Failed to parse command arguments: {err:?}");
-            std::process::exit(1);
-        }
-    };
+    use once_cell::sync::Lazy;
+    use tracing::warn;
 
-    #[cfg(feature = "jemalloc")]
-    start_jemalloc_metrics_loop();
-
-    setup_logging_and_tracing(
-        command.default_log_level(),
-        !matches.is_present("no-color"),
-        build_info,
-    )?;
-    let return_code: i32 = if let Err(err) = command.execute().await {
-        eprintln!("{} Command failed: {:?}\n", "✘".color(RED_COLOR), err);
-        1
-    } else {
-        0
-    };
-    quickwit_telemetry::send_telemetry_event(TelemetryEvent::EndCommand { return_code }).await;
-    telemetry_handle.terminate_telemetry().await;
-    global::shutdown_tracer_provider();
-    std::process::exit(return_code)
+    // we need that time reference to use an atomic and not a mutex for LAST_UNPARK
+    static TIME_REF: Lazy<Instant> = Lazy::new(Instant::now);
+
+    const ALLOWED_DELAY_MICROS: u64 = 5000;
+
+    thread_local!(static LAST_UNPARK: AtomicU64 = AtomicU64::new(0));
+    thread_local!(static MY_ID: AtomicU64 = AtomicU64::new(0));
+    static NEXT_ID: AtomicU64 = AtomicU64::new(1);
+
+    pub fn thread_unpark() {
+        LAST_UNPARK.with(|time| {
+            let now = Instant::now()
+                .checked_duration_since(*TIME_REF)
+                .unwrap_or_default();
+            time.store(now.as_micros() as u64, Ordering::Relaxed);
+        })
+    }
+
+    pub fn thread_park() {
+        LAST_UNPARK.with(|time| {
+            let now = Instant::now()
+                .checked_duration_since(*TIME_REF)
+                .unwrap_or_default();
+            let delta = now.as_micros() as u64 - time.load(Ordering::Relaxed);
+            if delta > ALLOWED_DELAY_MICROS {
+                let id = MY_ID.with(|my_id| {
+                    let id = my_id.load(Ordering::Relaxed);
+                    if id == 0 {
+                        let new_id = NEXT_ID.fetch_add(1, Ordering::Relaxed);
+                        my_id.store(new_id, Ordering::Relaxed);
+                        new_id
+                    } else {
+                        id
+                    }
+                });
+                warn!("Thread{id} wasn't parked for {delta}µs, is the runtime too busy?");
```

Comment:
We might want some mechanism to avoid logging this more than once in 30s or something.
or make it a counter?